auditlog/log-enricher/src/main/java/io/debezium/demos/auditing/enricher/TopologyProducer.java

```diff
@@ -0,0 +1,280 @@
+package io.debezium.demos.auditing.enricher;
+
+import java.time.Duration;
+import java.util.Optional;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Produces;
+import javax.json.Json;
+import javax.json.JsonObject;
+
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.StreamsBuilder;
+import org.apache.kafka.streams.Topology;
+import org.apache.kafka.streams.kstream.Materialized;
+import org.apache.kafka.streams.kstream.Transformer;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.PunctuationType;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.StoreBuilder;
+import org.apache.kafka.streams.state.Stores;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ApplicationScoped
+public class TopologyProducer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(TopologyProducer.class);
+
+    private static final Long BUFFER_OFFSETS_KEY = -1L;
+    private static final String STREAM_BUFFER_NAME = "stream-buffer-state-store";
+    private static final String STORE_NAME = "transaction-meta-data";
+
+
+    @ConfigProperty(name = "audit.context.data.topic")
+    String txContextDataTopic;
+
+    @ConfigProperty(name = "audit.vegetables.topic")
+    String vegetablesTopic;
+
+    @ConfigProperty(name = "audit.vegetables.enriched.topic")
+    String vegetablesEnrichedTopic;
+
+    @Produces
+    public Topology buildTopology() {
+        StreamsBuilder builder = new StreamsBuilder();
+
+        StoreBuilder<KeyValueStore<Long, JsonObject>> streamBufferStateStore =
+                Stores
+                    .keyValueStoreBuilder(
+                        Stores.persistentKeyValueStore(STREAM_BUFFER_NAME),
+                        new Serdes.LongSerde(),
+                        new JsonObjectSerde()
+                    )
+                    .withCachingDisabled();
+            builder.addStateStore(streamBufferStateStore);
+
+        builder.globalTable(txContextDataTopic, Materialized.as(STORE_NAME));
+
+        builder.<JsonObject, JsonObject>stream(vegetablesTopic)
+                // filter out any tombstones; for an audit log topic, time-based retention
+                // seems more reasonable than compaction
+                .filter((id, changeEvent) -> changeEvent != null)
+                // exclude snapshot events
+                .filter((id, changeEvent) -> !changeEvent.getString("op").equals("r"))
```

Comment:
Why are snapshot events ignored? I'm unable to see any reason why the existing code would fail to work for them.

Reply:
Good question, two things basically:

* As snapshot events are *read* events, I didn't see much reason to have them in an audit log kind of topic
* A bit more of an actual issue is that for snapshot events there won't be any corresponding TX record issued by the application typically; the snapshot TX is initiated by the Debezium connector, so the user would have to go, retrieve the TX id from one of the events and produce a matching record into the TX metadata topic; otherwise the snapshot events would block any other log-based events from being processed

So based on that I figured that's the best for the demo; in the post I'll discuss it, also showing options, e.g. one could want to propagate snapshot events un-enriched if that makes sense in a given use case.